ui/src/services/k8s/K8SLocalVolumeProvider.ts

```diff
@@ -0,0 +1,79 @@
+import { CoreV1Api, V1PersistentVolume } from '@kubernetes/client-node';
+import * as ApiK8s from '../k8s/api';
+import {
+  Metalk8sV1alpha1VolumeClient,
+  Result,
+} from './Metalk8sVolumeClient.generated';
+
+function isError<T>(result: Result<T>): result is { error: any } {
+  return (result as { error: any }).error !== undefined;
+}
+
+type LocalPersistentVolume = V1PersistentVolume & {
+  IP: string;
+  devicePath: string;
+};
+
+export interface K8SLocalVolumeAdapter {
+  listLocalPersistentVolumes(
+    serverName: string,
+  ): Promise<LocalPersistentVolume[]>;
+}
+
+export default class K8SLocalVolumeProvider implements K8SLocalVolumeAdapter {
+  volumeClient: Metalk8sV1alpha1VolumeClient;
+  k8sClient: CoreV1Api;
+  constructor(private url: string, private token: string) {
+    const { coreV1, customObjects } = ApiK8s.updateApiServerConfig(url, token);
+    this.volumeClient = new Metalk8sV1alpha1VolumeClient(customObjects);
+    this.k8sClient = coreV1;
+  }
+  public async listLocalPersistentVolumes(
+    serverName: string,
+  ): Promise<LocalPersistentVolume[]> {
+    try {
+      const nodes = await this.k8sClient.listNode();
+      const nodeIP = nodes.body.items
+        .find((node) => node.metadata.name === serverName)
+        ?.status.addresses.find((address) => address.type === 'InternalIP');
+
+      if (!nodeIP) {
+        throw new Error(`Failed to find IP for node ${serverName}`);
+      }
+
+      const volumes = await this.volumeClient.getMetalk8sV1alpha1VolumeList();
+
+      if (!isError(volumes)) {
+        const nodeVolumes = volumes.body.items.filter(
+          (volume) => volume.spec.nodeName === serverName,
+        );
+        const pv = await this.k8sClient.listPersistentVolume();
+
+        const localPv: LocalPersistentVolume[] = nodeVolumes.reduce(
+          (acc, item) => {
+            const isLocalPv = pv.body.items.find(
+              (p) => p.metadata.name === item.metadata['name'],
+            );
+            return [
+              ...acc,
+              {
+                ...isLocalPv,
+                IP: nodeIP.address,
+                devicePath: item.spec?.rawBlockDevice?.devicePath,
```

Comment:
Missing handling sparseloopdevice and lvm volumes

Reply:
You can refer to this documentation for data format of both https://metal-k8s.readthedocs.io/en/stable/operation/volume_management/volume_creation_deletion_cli.html